src/js/structs/Batch.js

```diff
@@ -55,6 +54,28 @@ class Batch {
    * @param {Object} item
    */
   add(item) {
+    let batchIndex = this.length;
+
+    // As the while is faster then the Array.prototype.forEach and this
+    // function can potentially be called more often we choose while here.
+    while (--batchIndex >= 0) {
+      let {path, value} = this[batchIndex] || {};
+      let hasEqualPaths = item && Array.isArray(path) &&
+        Array.isArray(item.path) && path.join() === item.path.join();
+      if (hasEqualPaths) {
+        // Last entry with same path holds same value,
+        // do not add redundant action
+        if (value === item.value) {
+          return;
+        }
+        // Remove previous if path is the same as current to minimize
+        // number of actions
+        if (batchIndex === this.length - 1) {
+          this.pop();
+        }
+      }
+    }
```

Comment:
Good points! I need to add some more tests it seems, however, I think it could still work with some tweaks.
For removing the previous event looking at the path should be enough, no need to check the value too.
If we only look back to the first equal path (which I guess was the intention) and check if the value is equal that should handle all those cases too.
Code would look something like this:
```js
let done = false;
while (!done && --batchIndex >= 0) {
  let {path, value} = this[batchIndex] || {};
  let hasEqualPaths = item && Array.isArray(path) &&
    Array.isArray(item.path) && path.join() === item.path.join();
  if (hasEqualPaths) {
    // Last entry with same path holds same value,
    // do not add redundant action
    if (value === item.value) {
      return;
    }
    // Remove previous if path is the same as current to minimize
    // number of actions
    if (batchIndex === this.length - 1) {
      this.pop();
    }
    // break out of while
    done = true;
  }
}
```